Adds WIP support for calling magic_setparam

pabs3 opened the pull request:
This is a first pass that works with my simple script and [test.jpg](https://bugs.launchpad.net/ubuntu/+source/file/+bug/1717991/+attachment/4952354/+files/test.jpg):

```
use File::LibMagic;
my %magic_params = (bytes => 857393);
my $magic = File::LibMagic->new(magic_params => \%magic_params);
my $info = $magic->info_from_filename('test.jpg');
print $info->{description};
```

Please review and let me know what mistakes I've made.

autarch commented on lib/File/LibMagic.pm:
```diff
@@ -90,6 +108,11 @@ sub new {
     # We need to call this even if $magic_paths is undef
     magic_load( $m, $magic_paths );
 
+    while ( ( my ( $name, $value ) ) = each(%$magic_params) ) {
```
I generally avoid `each` even though this particular case is safe.

autarch commented on lib/File/LibMagic.pm:
```diff
@@ -258,6 +281,11 @@ can't find any magic files at all.
+=item * magic_params
+
+This should be a hash containing one or more libmagic parameters.
+The keys should be the same as the `file --parameter` names.
```
We should list what's available explicitly.

autarch commented on lib/File/LibMagic.pm:
```diff
@@ -68,6 +75,16 @@ sub new {
     my $flags = MAGIC_NONE();
     my $magic_file;
+    my $magic_params    = ();
+    my %magic_param_map = (
```
This should be defined outside the sub since it doesn't change between invocations.

autarch commented on lib/File/LibMagic.pm:
```diff
@@ -78,6 +95,7 @@ sub new {
             if $p{follow_symlinks};
         $flags |= MAGIC_COMPRESS()
             if $p{uncompress};
+        $magic_params = $p{magic_params} if exists $p{magic_params};
```
I'm not sure there's any reason to make this a nested hashref instead of just allowing the parameters to be passed in to the constructor directly.

pabs3 commented:
It is my understanding that all non-trivial patches have copyright
attached, my patch seemed reasonably non-trivial so I added the fact
that I would have some copyright on File::LibMagic if the patch were to
be merged. If you don't think the changes are subject to copyright then
I'm fine to remove my name from the copyright holders.

I chose a nested hashref because the parameters to magic_setparam are
part of a separate namespace to parameters to the constructor. Right
now there are no conflicts but it is plausible that there could be in
the future. Either way it feels more correct. To be honest I think that
the follow_symlinks and uncompress options that are passed on to
magic_setflags should also be in a separate namespace %magic_flags.

I've fixed all the other issues you mentioned and a couple more.

I've also added support for passing an integer as the parameter name,
which is then directly passed to magic_setparam. This is so that if
libmagic upstream adds a new parameter, people can workaround
File::LibMagic not having it yet by just passing an integer.

I wasn't aware with the problems with %each, thanks for that tip.

BTW, do you think that errors from magic_setparam (and magic_setflags)
need to be handled?

-- 
bye,
pabs

https://wiki.debian.org/PaulWise

autarch commented:
> It is my understanding that all non-trivial patches have copyright attached, my patch seemed reasonably non-trivial so I added the fact that I would have some copyright on File::LibMagic if the patch were to be merged. If you don't think the changes are subject to copyright then I'm fine to remove my name from the copyright holders.

I'm fine with leaving it. It's just a bit odd, as I noted.

> I chose a nested hashref because the parameters to magic_setparam are part of a separate namespace to parameters to the constructor. Right now there are no conflicts but it is plausible that there could be in the future. Either way it feels more correct. To be honest I think that the follow_symlinks and uncompress options that are passed on to magic_setflags should also be in a separate namespace %magic_flags.

I disagree about it being more correct. The goal for this distribution is to provide a Perl-ish interface to the functionality provided by libmagic, _not_ to force people using the distro to understand all the nitty-gritty details of libmagic. If you're not familiar with the C lib, this distinction between what goes to `magic_setparam` and what doesn't isn't meaningful.

I'm not even sure it's meaningful in the library, except that one is a set of true/false flags (expressed via bit fiddling) and one is a set of integers. But in Perl we we can make everything a clearly named and documented parameter, without forcing users to understand the subtleties of the C lib interface.

> I've fixed all the other issues you mentioned and a couple more. I've also added support for passing an integer as the parameter name, which is then directly passed to magic_setparam. This is so that if libmagic upstream adds a new parameter, people can workaround File::LibMagic not having it yet by just passing an integer.

That's a good idea!

> BTW, do you think that errors from magic_setparam (and magic_setflags) need to be handled?

Yes, we should handle errors anywhere they can occur. The documentation doesn't say that those two functions can set an error, but looking at the source it does appear that it can. Gotta love old school C libraries and their docs.

pabs3 commented:
Point taken about the interface design. The future-compat integer param
feature will probably need to be different but I think I can probably
work out how to keep it within the interface.

Agreed re the error handling.

I'll rework the params patch and add a new one for magic_setflags errors.

-- 
bye,
pabs

https://wiki.debian.org/PaulWise

pabs3 commented:
I've updated the patch to use a more Perl-ish interface. I opted to
avoid any potential namespace issues by naming the parameters max_*
which also makes the Perl-side parameter names more self-explanatory.
I added max_future_compat as the support for future processing limits.
I updated the language in the docs to talk about processing limits.

-- 
bye,
pabs

https://wiki.debian.org/PaulWise

autarch commented:
Merged from the CLI. Thanks!
